### PR TITLE
optimize indexed color DDS conversion

### PR DESCRIPTION
**Pull request by mp3butcher**
Before : 3.6 ms
After    : 1.3 ms
(they are means over 1000 tries with ubuntu gcc 7.4.0 )

### CODE REVIEW

**Comment by Capostrophic** [comment]
If we're going there...
```cpp
unsigned char* convertedData = new unsigned char [sizeWithMipmaps * 4];
unsigned char* dest = convertedData;
unsigned char* index = imageData;
for (unsigned int i = 0; i < sizeWithMipmaps; i++)
{
    unsigned char* color = &palette[(*index++)*4];
    for (int j = 0; j < 4; j++)
        *dest++ = *color++;
}
```

File: src/osgPlugins/dds/ReaderWriterDDS.cpp [hunk 1]
```diff
@@ -1031,12 +1031,15 @@ osg::Image* ReadDDSFile(std::istream& _istream, bool flipDDSRead)
...
+        unsigned char *pimageData = imageData;
+        unsigned char *ppalet;
         for (unsigned int i = 0; i < sizeWithMipmaps; i++)
         {
-            convertedData[i * 4 + 0] = palette[imageData[i] * 4 + 0];
-            convertedData[i * 4 + 1] = palette[imageData[i] * 4 + 1];
-            convertedData[i * 4 + 2] = palette[imageData[i] * 4 + 2];
-            convertedData[i * 4 + 3] = palette[imageData[i] * 4 + 3];
+            ppalet = &palette[(*pimageData)*4];
+            memcpy(pconvertedData, ppalet, 4);
```
Comment (mp3butcher) [inline src/osgPlugins/dds/ReaderWriterDDS.cpp]: memcpy can be replace with 
`                   *((unsigned int *)pconvertedData)=*((unsigned int *)(ppalet));`
but it's not very pretty

Comment (Capostrophic) [reply src/osgPlugins/dds/ReaderWriterDDS.cpp]: Then I propose [this](https://github.com/Capostrophic/OpenSceneGraph/commit/028c7edf8bf8304fe448606ca898cb6dce788ee9).

**Comment by mp3butcher** [comment]
... considering adding -Wstrict-aliasing option might be a good thing...

**Comment by mp3butcher** [comment]
@robertosfield the last commit is left to your consideration

**Comment by openscenegraph** [comment]
This PR has got a bit silly with 7 different commits thrashing between different versions.  The last one looks like it might OK and is the most readable, has this one been profiled against the original version?  It seems like this last rev should be just submitted in it's own PR as I don't want all the intermediate commits merged as well.

**Comment by mp3butcher** [comment]
> This PR has got a bit silly with 7 different commits thrashing between different versions. The last one looks like it might OK and is the most readable, has this one been profiled against the original version? 

Indeed, profiler timings are between master implementation and this pr

> It seems like this last rev should be just submitted in it's own PR as I don't want all the intermediate commits merged as well.

You said the lastone https://github.com/openscenegraph/OpenSceneGraph/pull/822/commits/81bb7f2bf0a9c4be7742efe49353943666429d84 might be ok and readable, and now you says this commit should be object of an other pr: I don't follow you.

Should remove the last commit and squash all others in a single commit?

**Comment by openscenegraph** [comment]
I don't want the history of all your experiments.  The last incarnation of the code is what I want, this should be a minimal diff over the original code.

**Comment by mp3butcher** [comment]
ok i clarified this pr with 2 commits:
- first one use union to take care of type aliasing
- second use memcpy instead (not sure about the aliasing union trick)

**Pull request merged**